### ISSUE TITLE
Express rheology term as a function of viscous coefficients for EVP and rEVP

### ISSUE THREAD

**Issue by JFLemieux73** [CICE-Consortium/CICE#634]
The goal is to ease the implementation of new yield curves (e.g. Mohr-Coulomb) and flow rules (i.e. plastic potential).

**Comment by JFLemieux73** [comment]
@eclare108213  could you please have a look at what I coded so far?

git@github.com:JFLemieux73/CICE.git (branch = ViscCoeff). 

The only files that change are ice_dyn_evp and ice_dyn_shared. Look at the stress subroutine in evp. I know you would prefer to  keep the 'old' code as well but I would suggest we only keep the new code (in a sense the new code just has different variable names).

**Comment by phil-blain** [comment]
Direct link: https://github.com/CICE-Consortium/CICE/compare/master...JFLemieux73:ViscCoeff

**Comment by eclare108213** [comment]
@phil-blain thank you for the direct link!  Very helpful.
@JFLemieux73.  I agree, this is clean and would be messy keeping both versions.  I assume that this will not be BFB with the old code, so rather than keeping two versions, let's step this one through the QC process to make sure that the changes don't produce something unexpected, and merge it before you start to make a PR with Mohr-Coulomb modifications.  Let's also pay attention to computational cost - shouldn't be a big difference.  Finally, do changes need to be made to the documentation?  I don't think we have this level of detail in there, but we should check.  Sound okay?
(Lot's of "royal we's" in there :) )

**Comment by JFLemieux73** [comment]
Thanks @eclare108213. 

-I expect it to be BFB with respect to standard EVP and standard rEVP. 
-Do we have a process for assessing the computational cost?
-Let me have a look at the doc to see what needs to be changed.

**Comment by eclare108213** [comment]
> -I expect it to be BFB with respect to standard EVP and standard rEVP. 

Great - demonstrating that would make this even easier to merge in right away.

-Do we have a process for assessing the computational cost? 

There is a computational cost estimate based on the timing output, I think, as part of the tests suites.  @apcraig set it up and knows how it is supposed to work.

-Let me have a look at the doc to see what needs to be changed.

Thank you.

**Comment by JFLemieux73** [comment]
I find that arlx (and arlx1i) and brlx are kind of cryptic names for variables. Should I take this opportunity to change these? Any ideas? They are called alpha and beta in the Bouillon paper. 

alpha_revp, beta_revp (or alpha_evp and beta_evp as they are used for both evp and revp)?

**Comment by eclare108213** [comment]
I agree, they are very cryptic, but I actually prefer that to using spelled-out greek letters, which can stand for many different things.  For example, alpha is often used for albedo -- that's why I came up with arlx etc, but I don't remember why I chose the letters after 'a'!  I read a lucid description of what these parameters are in one of Sergei's papers -- e.g. one of them controls damping (or maybe it was convergence) for the momentum equation, the other one for stress, or something like that - I don't remember exactly.  Forming a name based on what the parameters actually do would be preferable, e.g. something like coef_damp_stress (but with fewer underscores so then coef_dampstress, or maybe damp_evpstress or similar/appropriate)?

**Comment by JFLemieux73** [comment]
Thanks Elizabeth...I will think about this a bit more.

**Comment by JFLemieux73** [comment]
Ok it is not BFB. The differences are very small (different order of operations?). The issue is with stressp. If I calc stressm and stress12 the new way (with eta) while stressp is calc the old way (with c0, c1) it is BFB. It is only with stressp calc with zeta and rep_prs that it is not BFB.

**Comment by eclare108213** [comment]
Not being BFB makes it even more important, in my opinion, to go through the QC process and merge the code now, before making additional changes for Mohr-Coulomb.  Since QC requires 5-year simulations, you could check the difference in timings (fingers crossed that the timers are actually working correctly).

**Comment by JFLemieux73** [comment]
Thanks Elizabeth. I am not ready anyway for Mohr-Coulomb. 

I will do a few PRs before that. I will do the QC 5-year simulations.

**Comment by JFLemieux73** [comment]
I am quite convinced that the small differences are due to changing the order of the operations. If I do 

stressp_1(i,j) = (stressp_1(i,j)*(c1-arlx1i*revp) + c1ne*divune*(c1+Ktens) - c1ne*Deltane*(c1-Ktens)) * denom1

instead of the standard approach

 stressp_1(i,j) = (stressp_1(i,j)*(c1-arlx1i*revp) + c1ne*(divune*(c1+Ktens) - Deltane*(c1-Ktens))) * denom1

it is NOT BFB even though this is exactly the same from a mathematical point of view.

**Comment by TillRasmussen** [comment]
The order of calculation can easily matter. One of the bugs that I found in the implementation of the 1d solver was the calculation of tinyarea:
      tmparea = dxt(iw) * dyt(iw) ! necessary to split calc of tinyarea. Otherwize not binary identical
      tinyarea =  puny * tmparea
and 
      tinyarea = puny *  dxt(iw) * dyt(iw)
is not bit for bit due to roundoffs. Both are equally right in a Mathematical sense.

**Comment by TillRasmussen** [comment]
> @TillRasmussen as you see I am reformulating a bit how the stresses are expressed so that new rheologies can be easily implemented. Should I give it a try with the 1devp or you prefer to do it yourself?

I can change the 1d evp when the 2d is accepted.

**Comment by JFLemieux73** [comment]
Thanks Till. I should do the PR soon. Here are the results of the end-to-end testing procedure.

387 measured results of 387 total results
259 of 387 tests PASSED
0 of 387 tests PENDING
0 of 387 tests MISSING data
128 of 387 tests FAILED

I am running the QC test right now.

**Comment by JFLemieux73** [comment]
The QC test has passed.

INFO:__main__:Running QC test on the following directories:
INFO:__main__:  /home/jfl001/data/eccc-ppp4/TEMPcice/base
INFO:__main__:  /home/jfl001/data/eccc-ppp4/TEMPcice/test
INFO:__main__:Number of files: 1825
INFO:__main__:2 Stage Test Passed
INFO:__main__:Quadratic Skill Test Passed for Northern Hemisphere
INFO:__main__:Quadratic Skill Test Passed for Southern Hemisphere

**Comment by JFLemieux73** [comment]
And for the test (new code)

Timer   1:     Total    5883.38 seconds
Timer   2:  TimeLoop    5882.98 seconds
Timer   3:  Dynamics    1183.10 seconds
Timer   4: Advection    1248.12 seconds
Timer   5:    Column    2442.08 seconds
Timer   6:    Thermo    2180.49 seconds
Timer   7: Shortwave     490.78 seconds
Timer   8:   Ridging     293.33 seconds
Timer   9: Floe size       0.00 seconds
Timer  10:  Coupling     752.44 seconds
Timer  11: ReadWrite     105.87 seconds
Timer  12:     Diags     186.42 seconds
Timer  13:   History      79.84 seconds
Timer  14:     Bound    3312.35 seconds
Timer  15:       BGC       0.00 seconds
Timer  16:   Forcing     636.61 seconds
Timer  17:    1d-evp       0.00 seconds
Timer  18:    2d-evp    1065.74 seconds

**Comment by JFLemieux73** [comment]
I also ran the QC test for the revised EVP:

INFO:__main__:Number of files: 1825
INFO:__main__:2 Stage Test Passed
INFO:__main__:Quadratic Skill Test Passed for Northern Hemisphere
INFO:__main__:Quadratic Skill Test Passed for Southern Hemisphere

**Comment by apcraig** [comment]
With regard to timing, I think the best we can say is that it's the same.  Would take several samples to see if anything was statistically different.  In fact, only the dynamics timer (and maybe a couple other timers that are "on" in the section that's changed) should be different.
